pkg/util/pki/renewaltime_test.go

```diff
@@ -87,7 +94,109 @@ func TestRenewalTime(t *testing.T) {
 		t.Run(n, func(t *testing.T) {
 			renewalTime := RenewalTime(s.notBefore, s.notAfter, s.renewBeforeOverride)
 			assert.Equal(t, s.expectedRenewalTime, renewalTime, fmt.Sprintf("Expected renewal time: %v got: %v", s.expectedRenewalTime, renewalTime))
+		})
+	}
+}
+
+func TestRenewBefore(t *testing.T) {
+	const duration = time.Hour * 3
+
+	type scenario struct {
+		renewBeforeOverride *string
+		expectedRenewBefore time.Duration
+	}
+
+	tests := map[string]scenario{
+		"spec.renewBefore is not set": {
+			renewBeforeOverride: nil,
+			expectedRenewBefore: time.Hour,
+		},
+		"spec.renewBefore is valid percentage": {
+			renewBeforeOverride: ptr.To("25%"),
+			expectedRenewBefore: 135 * time.Minute,
+		},
+		"spec.renewBefore is too large of a percentage so default is used": {
+			renewBeforeOverride: ptr.To("100%"),
+			expectedRenewBefore: time.Hour,
+		},
+		"spec.renewBefore is too small of a percentage so default is used": {
+			renewBeforeOverride: ptr.To("0%"),
+			expectedRenewBefore: time.Hour,
+		},
+		"spec.renewBefore is invalid percentage so default is used": {
+			renewBeforeOverride: ptr.To("xxx%"),
+			expectedRenewBefore: time.Hour,
+		},
+		"spec.renewBefore is empty so default is used": {
+			renewBeforeOverride: ptr.To(""),
+			expectedRenewBefore: time.Hour,
+		},
+		"spec.renewBefore is valid duration": {
+			renewBeforeOverride: ptr.To("1h"),
+			expectedRenewBefore: time.Hour,
+		},
+		"spec.renewBefore is invalid duration so default is used": {
+			renewBeforeOverride: ptr.To("4h"),
+			expectedRenewBefore: time.Hour,
+		},
+	}
+	for n, s := range tests {
+		t.Run(n, func(t *testing.T) {
+			renewBefore := RenewBefore(duration, s.renewBeforeOverride)
+			assert.Equal(t, s.expectedRenewBefore, renewBefore, fmt.Sprintf("Expected renewBefore time: %v got: %v", s.expectedRenewBefore, renewBefore))
+		})
+	}
+}
 
+func TestParseDurationOrPercent(t *testing.T) {
+	type scenario struct {
+		input            string
+		expectedDuration *time.Duration
+		expectedPercent  *int
+		expectedErr      error
+	}
+
+	tests := map[string]scenario{
+		"empty": {
+			input:       "",
+			expectedErr: fmt.Errorf(`time: invalid duration ""`),
+		},
+		"valid percent": {
+			input:           "25%",
+			expectedPercent: ptr.To(25),
+		},
+		"negative percent": {
+			input:       "-100%",
+			expectedErr: fmt.Errorf(`invalid percent: -100`),
+		},
+		"invalid percent": {
+			input:       "xxx%",
+			expectedErr: fmt.Errorf(`strconv.Atoi: parsing "xxx": invalid syntax`),
+		},
+		"floating point percent": {
+			input:       "4.5%",
+			expectedErr: fmt.Errorf(`strconv.Atoi: parsing "4.5": invalid syntax`),
+		},
+		"valid duration": {
+			input:            "1h",
+			expectedDuration: ptr.To(time.Hour),
+		},
+		"invalid duration": {
+			input:       "400d",
+			expectedErr: fmt.Errorf(`time: unknown unit "d" in duration "400d"`),
+		},
```

Comment:
What does the value "100" give as a result? Is this an error or is this parsed as an absolute duration?
Could you add that testcase?

Reply:
Values are parsed as percentages iff they end with a trailing `%`. Thus `100` is parsed as duration, which fails due to the missing unit:
```go
time: missing unit in duration "100"
```
I added a test case for this to make it explicit.